### ISSUE TITLE
[CdvPurchase] INFO: order(undefined) at the time of order place

### ISSUE THREAD

**Issue by RintuMondal06** [j3k0/cordova-plugin-purchase#1493]
### Project Info

* Plugin Version : ^13.9.0
* Ionic : @ionic/angular 5.9.4
* Cordova android: 12

```javascript <!-- syntax-highlighting:  paste your code below -->
// Consumable product register
   CdvPurchase.store.register({
	id: currentValue,
	type: CdvPurchase.ProductType.CONSUMABLE,
	platform:CdvPurchase.Platform.GOOGLE_PLAY,
    });


// event listeners
              CdvPurchase.store.when(my_product_code)
			.approved((transaction) =>transaction.verify())
			.verified((receipt) => receipt.finish())
			.finished((p) => {
			 // custom function called here to update the server
		});

// order code block
      CdvPurchase.store.order(my_product_code)
	.then((p) => {
	console.log('order then block', p);
	})
	.catch((e) => {
	console.log('error order in catch', e);
	});
```

### Output

Getting errors in order with error code `6777008` and the error message is `"Adapter not found or not ready (undefined)"`

**Comment by RintuMondal06** [comment]
> Have you initialized the store?
> 
> CdvPurchase.store.initialize();

@undergroundcreative yes, i called `CdvPurchase.store.initialize() ` before calling  `CdvPurchase.store.register()`!

**Comment by undergroundcreative** [comment]
Ok you need to register products and set callbacks before calling CdvPurchase.store.initialize()

See here for an example: https://purchase.cordova.fovea.cc/discover/micro-example

**Comment by RintuMondal06** [comment]
> Ok you need to register products and set callbacks before calling CdvPurchase.store.initialize()
> 
> See here for an example: https://purchase.cordova.fovea.cc/discover/micro-example

@undergroundcreative thanks for your help! I'm following the example from the doc but in my case `store.get()` is returning `undefined` while fetching the product. 

Below is my code:
```javascript <!-- syntax-highlighting:  paste your code below -->
     const product = CdvPurchase.store.get(
				PRODUCT_ID,
				CdvPurchase.Platform.GOOGLE_PLAY
			);

     product.getOffer()
		.order()
		.then((p) => {
			console.log('order then block', p);
		})
		.catch((e) => {
			console.log('error order in catch', e);
		});
```

**Comment by RintuMondal06** [comment]
This is the full code for the in-app-purchase.

```javascript <!-- syntax-highlighting:  paste your code below -->
// Consumable product register
   CdvPurchase.store.register({
	id: currentValue,
	type: CdvPurchase.ProductType.CONSUMABLE,
	platform:CdvPurchase.Platform.GOOGLE_PLAY,
    });


// event listeners
              CdvPurchase.store.when(my_product_code)
			.approved((transaction) =>transaction.verify())
			.verified((receipt) => receipt.finish())
			.finished((p) => {
			 // custom function called here to update the server
		});

// order code block
           const product = CdvPurchase.store.get(
				PRODUCT_ID,
				CdvPurchase.Platform.GOOGLE_PLAY
			);

     product.getOffer()
		.order()
		.then((p) => {
			console.log('order then block', p);
		})
		.catch((e) => {
			console.log('error order in catch', e);
		});
```

**Comment by undergroundcreative** [comment]
There's no sign of CdvPurchase.store.initialize(); in that code :)

What is your output now?

**Comment by RintuMondal06** [comment]
```javascript <!-- syntax-highlighting:  paste your code below -->
const product = CdvPurchase.store.get(
      'my_product_id',
      CdvPurchase.Platform.GOOGLE_PLAY
    );
    CdvPurchase.store
      .order(product.getOffer())
      .then((p: any) => {
        console.log('order then block', p);
      })
      .catch((e: any) => {
        console.log('error order in catch', e);
      });
```

`product` is undefined, that is my issue right now.

**Comment by AlexandrDegteari** [comment]
same problem I have, someone found solution?

**Comment by RubioN** [comment]
I have the same issue on my side, the function is returning undefined and some fields are missing in `products`.

**Comment by shahdeep1989** [comment]
@j3k0 @RintuMondal06 
I am facing same issue on iOS. Not tried for Android. Any solution for the same?

**Comment by j3k0** [comment]
If the product is undefined it could mean it isn't available for purchase. Your code needs to check if `store.get(...)` returns undefined and warn the user that the product isn't available at the moment.

Notice also that if done properly, the user shouldn't be able to call order for that product, since you're supposed to display the price and title as returned by the plugin. So if a product is not available the "buy" button should be grayed out or replaced by a warning that "this offer isn't available at the moment".

This is a use case that needs to be handled gracefully, as some users have in-app purchases disabled on their device.

To understand the reason why the product isn't available, please share some logs.

**Comment by shahdeep1989** [comment]
@j3k0 
store.get(...) returns undefined on iOS.
Also I checked again on App Store Connect and I can see status of Non-Renewing Subscriptions is "Approved". 
Why still I am not getting product details?

**Comment by j3k0** [comment]
>  Why still I am not getting product details?

Can you share debug logs?

**Comment by shahdeep1989** [comment]
@j3k0 

For android my in app purchase is working well. Getting issue on iOS. Sending log soon. 
Is it due to below reason?  Check screenshot.

![Uploading Screenshot 2024-01-12 at 12.28.35 PM.png…]()

**Comment by j3k0** [comment]
Yes, any banking, tax and paid agreements warnings will results in in-app purchases being disabled for the developer account.

**Comment by shahdeep1989** [comment]
@j3k0 All of the agreement and other setup has been done related to Tax, banking, etc. All set but still getting product details undefined on iOS, for Android it's working well.

Also follow the example but still getting undefined. Status of product on apple store is approved. And it was working previously. 
https://purchase.cordova.fovea.cc/discover/micro-example

Any one has any idea?

**Comment by shahdeep1989** [comment]
@j3k0 It's working now.

**Comment by j3k0** [comment]
So it was the missing tax form?

**Comment by shahdeep1989** [comment]
No it was one mistake in my code to validate platform. 

One more question:
store.when().verified(receipt => {

When I am checking receipt of testflight build on real device, I am not getting "receipt.sourceReceipt.transactions[0].nativePurchase".

Format of testflight build receipt has been changed?

**Comment by RintuMondal06** [comment]
@shahdeep1989 can you please help with Android implementation?  I can't find any issues in my current code. If you share your code here for reference it would be a great help for me. 
Thanks!

**Comment by JeremyCaltranz** [comment]
Hi, just leaving a helpful note here because I had this issue for weeks and finally figured out the issue for Android/Google Play.

CdvPurchase.store.order(yourOrderID); will always fail with an error "Product not registered: null". Even though the products are registered.

The proper usage appears to be:

CdvPurchase.store.get(yourOrderID).getOffer().order();

This second command works and triggers the Google Play Store to initiate the purchase.

I was able to connect my Android to my computer and then inspect the App via Chrome (Just put this in your chrome browser: chrome://inspect/#devices). Then you can run JS commands from the Chrome console to the App installed on your Android device.

**Comment by BigBalli** [comment]
`order(undefined)` means the offer object you're passing to `store.order()` is undefined. This almost always means `store.register()` and `store.initialize()` are being called in the wrong order, or `store.order()` is called before products finish loading.

The correct sequence:
```js
// 1. Register products
CdvPurchase.store.register({
  id: 'your_product',
  type: CdvPurchase.ProductType.CONSUMABLE,
  platform: CdvPurchase.Platform.GOOGLE_PLAY
});

// 2. Set up event handlers
CdvPurchase.store.when()
  .approved(t => t.verify())
  .verified(r => r.finish());

// 3. Initialize (this loads products from the store)
await CdvPurchase.store.initialize([CdvPurchase.Platform.GOOGLE_PLAY]);

// 4. Now get the product and offer
const product = CdvPurchase.store.get('your_product');
const offer = product?.getOffer();

// 5. Order
if (offer) await CdvPurchase.store.order(offer);
```

If `product` or `offer` is still undefined after `initialize()` resolves, the product ID doesn't match what's configured in Google Play Console, or the console setup isn't complete (products must be Active, app must be on at least a closed testing track, and the device account must be a license tester).
